File: backend/src/routes/achievements.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import { Router } from "express";
import { requireAuth, type AuthedRequest } from "../middleware/auth.js";
import { validateBody } from "../middleware/validate.js";
import { achievementCreate, achievementUpdate } from "../schemas/achievement.js";
import { getUserSupabaseClient } from "../supabaseClients.js";

export const achievementsRouter = Router();

achievementsRouter.use(requireAuth);

async function loadSkillIds(client: SupabaseClient, achievementId: string): Promise<string[]> {
  const { data } = await client
    .from("achievement_skills")
    .select("skill_id")
    .eq("achievement_id", achievementId);
  return (data ?? []).map((row) => row.skill_id as string);
}

async function attachSkillIds(
  achievement: Record<string, unknown>,
  client: SupabaseClient,
): Promise<Record<string, unknown>> {
  achievement.skill_ids = await loadSkillIds(client, achievement.id as string);
  return achievement;
}

function definedEntries(obj: Record<string, unknown>): Record<string, unknown> {
  return Object.fromEntries(Object.entries(obj).filter(([, v]) => v !== undefined));
}

achievementsRouter.get("", async (req, res) => {
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  const { data, error } = await client
    .from("achievements")
    .select("*")
    .order("created_at", { ascending: false });
  if (error) {
    res.status(400).json({ detail: error.message });
    return;
  }
  const withSkills = await Promise.all(data.map((a) => attachSkillIds(a, client)));
  res.json(withSkills);
});

achievementsRouter.post("", validateBody(achievementCreate), async (req, res) => {
  const { token, userId } = req as AuthedRequest;
  const client = getUserSupabaseClient(token);
  const { skill_ids: skillIds, ...rest } = req.body;

  const { data: created, error } = await client
    .from("achievements")
    .insert({ ...rest, user_id: userId })
    .select()
    .single();
  if (error) {
    res.status(400).json({ detail: error.message });
    return;
  }

  if (skillIds?.length) {
    const links = skillIds.map((skillId: string) => ({
      user_id: userId,
      achievement_id: created.id,
      skill_id: skillId,
    }));
    await client.from("achievement_skills").insert(links);
  }

  res.status(201).json(await attachSkillIds(created, client));
});

async function fetchAchievement(client: SupabaseClient, achievementId: string) {
  const { data, error } = await client
    .from("achievements")
    .select("*")
    .eq("id", achievementId);
  if (error || !data?.length) {
    return null;
  }
  return attachSkillIds(data[0], client);
}

achievementsRouter.get("/:achievementId", async (req, res) => {
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  const achievement = await fetchAchievement(client, req.params.achievementId);
  if (!achievement) {
    res.status(404).json({ detail: "Achievement not found" });
    return;
  }
  res.json(achievement);
});

achievementsRouter.patch("/:achievementId", validateBody(achievementUpdate), async (req, res) => {
  const { token, userId } = req as AuthedRequest;
  const client = getUserSupabaseClient(token);
  const achievementId = req.params.achievementId as string;
  const { skill_ids: skillIds, ...rest } = req.body;
  const fields = definedEntries(rest);

  if (Object.keys(fields).length > 0) {
    const { data, error } = await client
      .from("achievements")
      .update(fields)
      .eq("id", achievementId)
      .select();
    if (error || !data?.length) {
      res.status(404).json({ detail: "Achievement not found" });
      return;
    }
  }

  if (skillIds !== undefined) {
    await client.from("achievement_skills").delete().eq("achievement_id", achievementId);
    if (skillIds.length) {
      const links = skillIds.map((skillId: string) => ({
        user_id: userId,
        achievement_id: achievementId,
        skill_id: skillId,
      }));
      await client.from("achievement_skills").insert(links);
    }
  }

  const achievement = await fetchAchievement(client, achievementId);
  if (!achievement) {
    res.status(404).json({ detail: "Achievement not found" });
    return;
  }
  res.json(achievement);
});

achievementsRouter.delete("/:achievementId", async (req, res) => {
  const client = getUserSupabaseClient((req as AuthedRequest).token);
  await client.from("achievements").delete().eq("id", req.params.achievementId);
  res.status(204).send();
});

File: backend/app/routers/achievements.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from supabase import Client

from ..deps import get_current_user_id, get_user_supabase_client
from ..schemas import AchievementCreate, AchievementOut, AchievementUpdate

router = APIRouter(prefix="/achievements", tags=["achievements"])


def _load_skill_ids(client: Client, achievement_id: str) -> list[str]:
    response = (
        client.table("achievement_skills")
        .select("skill_id")
        .eq("achievement_id", achievement_id)
        .execute()
    )
    return [row["skill_id"] for row in response.data]


def _attach_skill_ids(achievement: dict, client: Client) -> dict:
    achievement["skill_ids"] = _load_skill_ids(client, achievement["id"])
    return achievement


@router.get("", response_model=list[AchievementOut])
def list_achievements(client: Client = Depends(get_user_supabase_client)) -> list[dict]:
    response = (
        client.table("achievements")
        .select("*")
        .order("created_at", desc=True)
        .execute()
    )
    return [_attach_skill_ids(a, client) for a in response.data]


@router.post("", response_model=AchievementOut, status_code=status.HTTP_201_CREATED)
def create_achievement(
    achievement: AchievementCreate,
    client: Client = Depends(get_user_supabase_client),
    user_id: str = Depends(get_current_user_id),
) -> dict:
    data = achievement.model_dump(mode="json", exclude={"skill_ids"})
    data["user_id"] = user_id
    response = client.table("achievements").insert(data).execute()
    created = response.data[0]

    if achievement.skill_ids:
        links = [
            {
                "user_id": user_id,
                "achievement_id": created["id"],
                "skill_id": str(skill_id),
            }
            for skill_id in achievement.skill_ids
        ]
        client.table("achievement_skills").insert(links).execute()

    return _attach_skill_ids(created, client)


@router.get("/{achievement_id}", response_model=AchievementOut)
def get_achievement(
    achievement_id: UUID, client: Client = Depends(get_user_supabase_client)
) -> dict:
    response = (
        client.table("achievements").select("*").eq("id", str(achievement_id)).execute()
    )
    if not response.data:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Achievement not found")
    return _attach_skill_ids(response.data[0], client)


@router.patch("/{achievement_id}", response_model=AchievementOut)
def update_achievement(
    achievement_id: UUID,
    achievement: AchievementUpdate,
    client: Client = Depends(get_user_supabase_client),
    user_id: str = Depends(get_current_user_id),
) -> dict:
    payload = achievement.model_dump(
        mode="json", exclude_unset=True, exclude={"skill_ids"}
    )
    if payload:
        response = (
            client.table("achievements")
            .update(payload)
            .eq("id", str(achievement_id))
            .execute()
        )
        if not response.data:
            raise HTTPException(status.HTTP_404_NOT_FOUND, "Achievement not found")

    if achievement.skill_ids is not None:
        client.table("achievement_skills").delete().eq(
            "achievement_id", str(achievement_id)
        ).execute()
        if achievement.skill_ids:
            links = [
                {
                    "user_id": user_id,
                    "achievement_id": str(achievement_id),
                    "skill_id": str(skill_id),
                }
                for skill_id in achievement.skill_ids
            ]
            client.table("achievement_skills").insert(links).execute()

    return get_achievement(achievement_id, client)


@router.delete("/{achievement_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_achievement(
    achievement_id: UUID, client: Client = Depends(get_user_supabase_client)
) -> None:
    client.table("achievements").delete().eq("id", str(achievement_id)).execute()
